website/pages/docs/internals/filesystem.mdx

```diff
@@ -0,0 +1,428 @@
+[filesystem isolation capability].
+
+### `image` isolation
+
+Task drivers like `docker` or `qemu` use `image` isolation, where the task
+driver isolates task filesystems as machine images. These filesystems are
+owned by the task driver's external process and not by Nomad itself. These
+filesystems will not typically be found anywhere in the allocation working
+directory. For example, Docker containers will have their overlay filesystem
+unpacked to `/var/run/docker/containerd/:container_id` by default.
+
+Nomad will provide the `NOMAD_ALLOC_DIR`, `NOMAD_TASK_DIR`, and
+`NOMAD_SECRETS_DIR` to tasks with `image` isolation, typically by
+bind-mounting them to the task driver's filesystem.
+
+You can see an example of `image` isolation by running the following minimal
+job:
+
+```hcl
+job "example" {
+  datacenters = ["dc1"]
+
+  task "task1" {
+    driver = "docker"
+
+    config {
+      image = "redis:6.0"
+    }
+  }
+}
+```
+
+If you look at the allocation working directory from the host, you'll see a
+minimal filesystem tree:
+
+```shell-session
+.
+├── alloc
+│   ├── data
+│   ├── logs
+│   │   ├── task1.stderr.0
+│   │   └── task1.stdout.0
+│   └── tmp
+└── task1
+    ├── local
+    ├── secrets
+    └── tmp
+```
+
+The `nomad alloc fs` command shows the same bare directory tree:
+
+```shell-session
+$ nomad alloc fs b0686b27
+Mode        Size     Modified Time         Name
+drwxrwxrwx  4.0 KiB  2020-10-27T18:51:54Z  alloc/
+drwxrwxrwx  4.0 KiB  2020-10-27T18:51:54Z  task1/
+
+$ nomad alloc fs b0686b27 task1
+Mode         Size     Modified Time         Name
+drwxrwxrwx   4.0 KiB  2020-10-27T18:51:54Z  local/
+drwxrwxrwx   60 B     2020-10-27T18:51:54Z  secrets/
+dtrwxrwxrwx  4.0 KiB  2020-10-27T18:51:54Z  tmp/
+
+$ nomad alloc fs b0686b27 task1/local
+Mode  Size  Modified Time  Name
+```
+
+If you inspect the Docker container that's created, you'll see three
+directories bind-mounted into the container:
+
+```shell-session
+$ docker inspect 32e | jq '.[0].HostConfig.Binds'
+[
+  "/var/nomad/alloc/b0686b27-8af3-8252-028f-af485c81a8b3/alloc:/alloc",
+  "/var/nomad/alloc/b0686b27-8af3-8252-028f-af485c81a8b3/task1/local:/local",
+  "/var/nomad/alloc/b0686b27-8af3-8252-028f-af485c81a8b3/task1/secrets:/secrets"
+]
+```
+
+The root filesystem inside the container can see these three mounts, along
+with the rest of the container filesystem:
+
+```shell-session
+$ docker exec -it 32e /bin/sh
+# ls /
+alloc  boot  dev  home  lib64  media  opt   root  sbin     srv  tmp  var
+bin    data  etc  lib   local  mnt    proc  run   secrets  sys  usr
+```
+
+Note that because the three directories are bind-mounted into the container
+filesystem, nothing written outside those three directories elsewhere in the
+allocation working directory will be accessible inside the container. This
+means templates, artifacts, and dispatch payloads for tasks with `image`
+isolation must be written into the `NOMAD_ALLOC_DIR`, `NOMAD_TASK_DIR`, or
+`NOMAD_SECRETS_DIR`.
```

Comment:
@angrycub do you think this section could use an example of how to re-mount that you showed us in https://github.com/hashicorp/nomad/issues/8962#issuecomment-699098754 or would that be too confusing here?

Reply:
It's probably one of the most common questions that pops up once someone starts templating.  I think there could be good value for more detail here about that.  Also, I wonder if we have the Tabs component in this framework.  Could let us slide in some more examples w/o making the page forever long.

Reply:
Ok, I'll update with an example and we can judge the length and things like tabs once that's up.

Reply:
Added that example in 054919d